### ISSUE TITLE
Missing variables in GSI ncdiag files cause inconsistent format in IODA-converted files

### ISSUE THREAD

**Issue by svahl991** [JCSDA-internal/ioda-converters#221]
Our group has been performing MPAS model experiments using the 1-month of cycling data provided in the ncdiag files at http://data.jcsda.org/gsidiags'+yyyymmdd+'.tar.

We take these ncdiag files and use the ''proc_gsi_ncdiag.py" converter script followed by the "combine_conv.py" to create a IODA file containing all the radiosonde conventional obs (t, q, u, and v) for each time period, that we then use for our experiments.

We noticed that that for one single time period in the entire month of data we were missing the specific_humidity (q) variable in the converted & combined file. To be precise, the problem is not that there are missing values in in the specific_humidity variable, but that the variable is missing entirely from the file. This causes problems for our post-processing scripts that are equipped to handle missing values in a variable, but expect the variable to be present in the file and populated with missing values. 

The time period with the issue is the '6Z' window of April 30, 2018. After some detective work (described in detail below), we found that this situation came about because the source 'conv_q' ncdiag file for that time period contained no specific humidity information at all, and it appears to be the only time period in the entire month where this occurs.

Since this issue appears to be caused not by an error, but simply because of missing data, somewhere the scripts that process these data are going to need to handle this situation.

The question is which scripts should deal with it. Should our post-processing script need to test that all the same variables are present in all the converted/combined files? Or should the issue be dealt with "upstream", so that the files that come out of the convert and combine process are uniform in format? The latter seems more desirable, but maybe more difficult?

I'm interested to hear your thoughts @CoryMartin-NOAA about the difficulty of solving this issue in the scripts that perform the convert & combine process.

As mentioned earlier, I'm including below the text of an email I wrote which contains all the details on the research I've done on this issue, in case it is helpful for understanding:

=======================
JJ, Junmei,

I've been looking into the case of the missing specific humidity data in the 2018043006 time window of our 1-month data set. I think I have figured out what is happening.

I started my investigation by looking at the source GSI ncdiag files. I thought maybe one of the ncdiag files would be missing in the '06Z' directory. But that was not the case. All the same files were present in that directory as all the others.

So then I looked to see if I could figure out how the 6Z 'conv_q' ncdiag file was different from the other 'conv_q' files in the other directories. I used ncdump to examine the values present in the Observation_Type variable in that file. There are about 7 or 8 different numeric values present in the Observation_Type vector of most of the files. But I noticed that the value '120' was present in all the files I examined except for the 6Z file. I noticed that the quantity of '120' values varied a lot between files, but the 6Z file was the only one where it was not present at all.

Then I looked at the netCDF files that the GSI ioda-converter creates from the ncdiag files. As you may know, the GSI ioda-converter produces a separate IODA file for each observation variable. Sure enough, when I looked at the 'sondes_q_obs' file created by the converter for various directories, the number of observations in that file contained corresponded to the number of '120' values I saw in the original source file. And in the 06Z folder with the problem, the 'sondes_q_obs' file was missing completely. Since the source file had no '120' values, the file didn't get created at all.

The last step of the conversion process is to run another python script that combines the separate IODA variable files output by the converter into a single IODA file containing all the variables. This script looks at the locations present in each variable file and "matches up" the different observation values for each location. If a variable value is missing for a location, then a Missing Value is assigned for that location.

So as long as the 'sondes_q_obs' file is present, even if it only has one value in it, the process works and the "specific_humidity@ObsValue" variable gets created in the combined file, even though it may be mostly populated with missing values. But when the entire file is missing, two things go wrong: 1) Our script that combines the files doesn't work because the 'sondes_q_obs' file is missing, and 2) Even if you manually create a combined file with the variables you do have, that combined file does not contain a specific_humidity@ObsValue variable.

So, bottom line, I think what we are seeing is not really an error with the source data; I think this is something that might occur from time to time. So I think it will need to be dealt with somewhere along the conversion process, or in the code that reads the converted files. I'm not sure where the proper place to deal with it is.

**Comment by CoryMartin-NOAA** [comment]
To clarify, are there input radiosonde q observations valid at that analysis cycle or are there no input observations? It seems weird that it would only be an issue for one cycle. Generally speaking, I do not like the combining of the variables (T,P,Q,UV) into one file as it can cause this sort of issue. The way GSI is set up is it only has the fields it needs in the routines to generate H(x) and this leads to missing values in GeoVaLs too (see the issue @MarekWlasak was having as well as UFO #660.

Since you are running MPAS experiments and have no need for the GSI H(x), this illustrates how the GSI workflow is really only a temporary solution and more work should be done on direct conversion from the input observation formats to IODA.

**Comment by svahl991** [comment]
I believe that there are no input radiosonde q observations for that cycle. When I examined the ncdiag 'conv_q' file for that cycle, there were no values of '120' present in the "Observation_Type" vector, and that appeared to be the only 'conv_q' file for all the cycles where that was true. I am assuming that the '120' value corresponds to the radiosonde q value? I could be wrong. I've never worked with the ncdiag files before. My email that I included at the end of my original post describes my whole investigation and thinking, so you can look that over and tell me if I've made a mistake somewhere.

My understanding is that the reason the scientists like the combined files is largely for plotting and analysis. If all the radiosonde obs for a profile are not combined into a single, uniform set of locations it is difficult to understand and analyze the data. Perhaps @jjguerrette can elaborate more?

**Comment by jjguerrette** [comment]
As far as postprocessing, I think either combined or individual files for a single ObsSpace can be handled.

Our group is trying to get the discussion started, not to prescribe a solution at this point.  The leading driver of the workflow is getting obs into IODA/UFO.  We are trying to get the post-processing to work with whatever that upstream procedure looks like.

To be honest I have not given much thought on the separate/togetherness of different obs variables in IODA-formatted files.  I suppose it makes collocation easier to have multiple variables in the same ObsSpace.  That does not mean they absolutely need to be in the same file, but the record numbers need to be consistent.  Whether the files are together or separate, what is the appropriate action when all q values are missing? 
Three possible options:
- tell the workflow manager to make the appropriate yaml changes (remove/do not add that variable or ObsSpace)
- generate an empty vector for the missing variable in the IODA formatted file
- do nothing and have downstream applications (DA and postprocessing) handle the situation individually

Is it premature to have this discussion?

**Comment by rhoneyager** [comment]
This issue has been inactive for a while. I have placed it in the AOP20 icebox for discussion in April. If resolved already, please close.

**Comment by jjguerrette** [comment]
This is probably a workflow problem and not an `ioda-converters` problem.  The workflow needs to decipher which variables are available in the obs file and then compile the list of `simulated variables`.  Closing.  Might be important to consider in EWOK.
